etk-asm/src/asm.rs

```diff
@@ -244,107 +231,127 @@ impl Assembler {
+        for op in self.ready.iter() {
+            if let RawOp::Op(ref op) = op {
+                match op
+                    .clone()
+                    .concretize((&self.declared_labels, &self.declared_macros).into())
+                {
+                    Ok(cop) => cop.assemble(&mut output),
+                    Err(ops::Error::ContextIncomplete {
+                        source: UnknownLabel { label: _label, .. },
+                    }) => {
+                        let undeclared_names: Vec<_> = self
+                            .undeclared_labels
+                            .iter()
+                            .map(|PendingLabel { label, .. }| label.clone())
+                            .collect();
+                        return error::UndeclaredLabels {
+                            labels: undeclared_names,
                         }
-                        _ => unreachable!(),
+                        .fail();
+                    }
+                    Err(ops::Error::ContextIncomplete {
+                        source: UnknownMacro { name, .. },
+                    }) => {
+                        return error::UndeclaredInstructionMacro { name }.fail();
                     }
+                    Err(ops::Error::ContextIncomplete {
+                        source: UndefinedVariable { name, .. },
+                    }) => {
+                        return error::UndeclaredVariableMacro { var: name }.fail();
+                    }
+
+                    Err(_) => unreachable!("all ops should be concretizable"),
                 }
-                // bug: if a variable is used when it isn't available, e.g. push1 $size
-                _ => unreachable!(),
-            };
+            } else if let RawOp::Raw(raw) = op {
+                output.extend(raw);
+            }
         }
 
-        if !self.ready.is_empty() {
-            panic!("not all assembled bytecode has been taken");
+        Ok(output)
+    }
+
+    /// Check if the input sequence is complete. Returns any errors that
+    /// may remain.
+    fn finish(&mut self) -> Result<(), Error> {
+        if !self.undeclared_labels.is_empty() {
+            return error::UndeclaredLabels {
+                labels: self
+                    .undeclared_labels
+                    .iter()
+                    .map(|l| l.label.to_owned())
+                    .collect::<Vec<String>>(),
+            }
+            .fail();
         }
 
         Ok(())
     }
 
-    /// Collect any assembled instructions that are ready to be output.
-    pub fn take(&mut self) -> Vec<u8> {
-        std::mem::take(&mut self.ready)
-    }
-
     /// Feed instructions into the `Assembler`.
     ///
-    /// Returns the number of bytes that can be collected with [`Assembler::take`].
-    pub fn push_all<I, O>(&mut self, ops: I) -> Result<usize, Error>
+    /// Returns the code of the assembled program.
+    pub fn assemble<I, O>(&mut self, ops: I) -> Result<Vec<u8>, Error>
```

Comment:
I think at this point we can just use a slice instead of an iterator.

```suggestion
    pub fn assemble<O>(&mut self, ops: &[O]) -> Result<Vec<u8>, Error>
```